File: code/classes/protocol.ts
// Protocol in TypeScript (using interface)
interface Drawable {
    draw(): void;
}

class Circle implements Drawable {
    draw(): void {
        console.log("Drawing a circle");
    }
}

function drawShape(shape: Drawable) {
    shape.draw();
}

const circle: Circle = new Circle();
drawShape(circle);

File: code/classes/protocol.java
// Protocol in Java (using interface)
interface Drawable {
    void draw();
}

class Circle implements Drawable {
    public void draw() {
        System.out.println("Drawing a circle");
    }
}

public class Main {
    public static void main(String[] args) {
        Drawable drawable = new Circle();
        drawable.draw();
    }
}